cachelib/shm/FileShmSegment.cpp

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cachelib/shm/FileShmSegment.h"
+
+#include <fcntl.h>
+#include <folly/logging/xlog.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include "cachelib/common/Utils.h"
+
+namespace facebook {
+namespace cachelib {
+
+constexpr static mode_t kRWMode = 0666;
+typedef struct stat stat_t;
+
+namespace detail {
+
+// TODO(SHM_FILE): move those *Impl functions to common file, there are copied
+// from PosixShmSegment.cpp
+static int openImpl(const char* name, int flags) {
+  const int fd = open(name, flags);
+
+  if (fd != -1) {
+    return fd;
+  }
+
+  switch (errno) {
+  case EEXIST:
+  case EMFILE:
+  case ENFILE:
+  case EACCES:
+    util::throwSystemError(errno);
+    break;
+  case ENAMETOOLONG:
+  case EINVAL:
+    util::throwSystemError(errno, "Invalid segment name");
+    break;
+  case ENOENT:
+    if (!(flags & O_CREAT)) {
+      util::throwSystemError(errno);
+    } else {
+      XDCHECK(false);
+      // FIXME: posix says that ENOENT is thrown only when O_CREAT
+      // is not set. However, it seems to be set even when O_CREAT
+      // was set and the parent of path name does not exist.
+      util::throwSystemError(errno, "Invalid errno");
+    }
+    break;
+  default:
+    XDCHECK(false);
+    util::throwSystemError(errno, "Invalid errno");
+  }
+  return kInvalidFD;
+}
+
+static void unlinkImpl(const char* const name) {
+  const int ret = unlink(name);
+  if (ret == 0) {
+    return;
+  }
+
+  switch (errno) {
+  case ENOENT:
+  case EACCES:
+    util::throwSystemError(errno);
+    break;
+  case ENAMETOOLONG:
+  case EINVAL:
+    util::throwSystemError(errno, "Invalid segment name");
+    break;
+  default:
+    XDCHECK(false);
+    util::throwSystemError(errno, "Invalid errno");
+  }
+}
+
+static void ftruncateImpl(int fd, size_t size) {
+  const int ret = ftruncate(fd, size);
+  if (ret == 0) {
+    return;
+  }
+  switch (errno) {
```

Comment:
This switch statement is repeated several times throughout this file. It will be better to write a function to do this error analysis and call it instead of copy/pasting the code.